### PR TITLE
Adds poetry support

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

Adds Poetry Support for Python projects.

I've closed previous PR due to significant noise - this addresses all the comments and nits. It should be easier to browse as well. Closed PR link is: https://github.com/fossas/spectrometer/pull/284
 
## Acceptance criteria

_What is this PR trying to accomplish?_

* User can get direct dependencies with `pyproject.toml`
* User can get direct and deep dependencies with `pyproject.toml` and `poetry.lock`
* User can use different package sources (git, url, etc.) 

## Testing plan

_How did you validate that this PR works? How did you check that the acceptance criteria were fulfilled?_

* I wrote unit tests, as well as manually tested against example projects with `fossa analyze --output` and `poetry show --tree --no-dev`
* I have performed end-to-end test with core

_This section should list reproduction steps that a reviewer can run through (and provide any needed test cases)._

* You can clone `https://github.com/meghfossa/example-projects` repository, it has example python poetry projects, or other poetry projects that you can try. 

## Risks

_Highlight any areas that you're unsure of, or want reviewers to pay particular attention to._

* Case handling of package references within poetry lock, but it is aptly documented. 

## References

_List any referenced GitHub issues. If PR references tickets in other systems (e.g. Zendesk), they should probably be mirrored as a GitHub Issue._

This closes fossas/team-analysis#549

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] I linted and formatted (via `haskell-language-server`) any haskell files I touched in this PR.
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Just one more required change, plus a few nits.

File: src/Effect/Grapher.hs
Comment (meghfossa) [reply src/Effect/Grapher.hs:60]: I've added this as techdebt issue in backlog: https://github.com/fossas/spectrometer/issues/302.

File: src/Strategy/Python/Poetry.hs
Comment (meghfossa) [inline src/Strategy/Python/Poetry.hs:83]: I tried using `$>` operator to simulate `as $> bs` ~ `do as; pure bs`. But ran into type issue.

Comment (scruffystuffs) [reply src/Strategy/Python/Poetry.hs:83]: I had no problem, left a suggestion that built for me.

File: src/Strategy/Python/Poetry/Common.hs
Comment (scruffystuffs) [inline src/Strategy/Python/Poetry/Common.hs:177]: [optional nit]: Since you unconditionally wrap in a `Just` here, the return type can probably be `a` instead of `Maybe a`.  You could move the `Just` constructor to the caller, and then this function will do less (often good for code organization)

**Review by scruffystuffs — APPROVED** [review]
I'd wait for @liftM to approve as well, but this is ready to go.  I left a few more nits, but nothing blocking at this point.

File: src/Strategy/Python/Poetry/Common.hs
Comment (liftM) [reply src/Strategy/Python/Poetry/Common.hs:53]: Really? Why do we prefer the less general one? Is there precedent for this elsewhere in the codebase?

Comment (scruffystuffs) [reply src/Strategy/Python/Poetry/Common.hs:53]: It's usually a better idea to tell the reader "I'm working with a list", rather than "I'm working with *some* functor, and it doesn't matter which one".  It's the same reason I'd rather see `Just`, instead of `pure`, although `Maybe` implements `Applicative` with `pure = Just` ([source here](https://hackage.haskell.org/package/base-4.15.0.0/docs/src/GHC-Base.html#line-1015))

File: src/Strategy/Python/Poetry.hs
Comment (liftM) [inline src/Strategy/Python/Poetry.hs:171]: I don't think I ever got an answer to this - why are you doing `buildPackageNameGraph` separately from `buildGraphWithLock`?

File: src/Strategy/Python/Poetry/Common.hs
Comment (liftM) [inline src/Strategy/Python/Poetry/Common.hs:1]: Why not combine this with `Strategy.Python.Poetry`?

**Review by liftM — APPROVED** [review]
Generally looks good. I didn't take a close look at the test correctness, since I looked in the last PR and they don't seem to have materially changed. Please let me know if these tests have changed significantly.

The most material piece of feedback here is the organization of `buildPackageNameGraph`. Please address this comment before merging (i.e. please either answer or refactor and check in with me before merging - feel free to ping me on Slack), but there's nothing in here I see that's a blocker.

File: src/Strategy/Python/Poetry.hs
Comment (meghfossa) [reply src/Strategy/Python/Poetry.hs:171]: There is _no single_ reason. The reason I structured in this manner was to provide some delineation between relationship (edges, and etc), and labelling of dependency. It also made logic more traceable/managable at the time. Happy to refractor here, if desired.

Comment (liftM) [reply src/Strategy/Python/Poetry.hs:169]: Oh, I see. In that case, I'd suggest commenting this intention explicitly.

File: src/Strategy/Python/Poetry.hs
Comment (liftM) [inline src/Strategy/Python/Poetry.hs:168]:
```suggestion
-- | Using a Poetry lockfile, build the graph of packages.
-- The resulting graph contains edges, but does not distinguish between direct and deep dependencies, since `poetry.lock` does not indicate which dependencies are direct.
```

File: src/Strategy/Python/Poetry.hs
Comment (liftM) [inline src/Strategy/Python/Poetry.hs]: Given your stated intentions, I would move this into `buildPackageNameGraph` such that `buildPackageNameGraph :: [PoetryLockPackage] -> Graphing Dependency` because:

1. This logic is only based on lockfile information, and doesn't require the pyproject.toml, so it's clearer that the intention of `buildPackageNameGraph` is "construct a graph from `poetry.lock`'s information".
2. This makes it clearer that `buildGraphWithLock` is just for setting direct dependencies using `pyproject.toml`.

I might also:

1. Rename `buildPackageNameGraph` to `graphFromLockfile :: [PoetryLockPackage] -> Graphing Dependency`.
2. Rename `buildGraphWithLock` to `setGraphDirectsFromPyproject :: Graphing Dependency -> PyProject -> Graphing Dependency`.
3. Call these two functions in `analyze` as appropriate.

File: src/Strategy/Python/Poetry.hs
Comment (liftM) [inline src/Strategy/Python/Poetry.hs:120]:
```suggestion
-- | Use a `pyproject.toml` to set the direct dependencies of a graph created from `poetry.lock`.
```

**Review by liftM — APPROVED** [review]
The knowledge of "`poetry.lock` contains deeps and edges, but not directs + `pyproject.toml` contains directs but no deeps or edges" is useful for implementers - I might add these as comments somewhere in `analyze`.

**Review by liftM — APPROVED** [review]

**Pull request merged**
